src/wasm/mod.rs

```diff
@@ -38,61 +47,91 @@ impl Server {
     #[wasm_bindgen(constructor)]
     pub fn new(disable_folding: bool) -> Server {
         Server {
-            handler: Handler::new(disable_folding),
+            handler: Rc::new(RefCell::new(Handler::new(
+                disable_folding,
+            ))),
+            callbacks: Callbacks::default(),
         }
     }
 
-    pub fn process(&mut self, msg: String) -> ServerResponse {
-        let mut lines = msg.lines();
+    pub fn register_buckets_callback(&mut self, f: Function) {
+        self.callbacks.register_buckets_callback(f);
+    }
 
-        if lines.clone().count() > 2 {
-            // Skip content length and spacer
-            lines.next();
-            lines.next();
+    pub fn get_buckets(&mut self) -> Promise {
+        let callbacks = self.callbacks.clone();
 
-            let mut content = String::new();
+        future_to_promise(async move {
+            let result = callbacks.get_buckets().await;
 
-            for line in lines {
-                content = content.add(line);
+            match result {
+                Ok(v) => Ok(JsValue::from(v.join(","))),
+                Err(e) => Err(JsValue::from(e)),
             }
+        })
+    }
+
+    pub fn process(&mut self, msg: String) -> Promise {
+        let handler = self.handler.clone();
+        let callbacks = self.callbacks.clone();
+
+        future_to_promise(async move {
+            let lines = msg.lines();
+
+            if lines.clone().count() > 2 {
```

Comment:
Maybe add `// Skip content length and spacer` back to be clear what it's skipping.